File: src/python/tools/natural_language.py
"""
Natural Language Processing tool for Airtable MCP Python
"""

import logging
import json
from typing import Any, Dict, Optional
from ..nlp import (
    NaturalLanguageProcessor,
    NLPConfig,
    NaturalLanguageQuery,
    ProcessedQuery,
    NaturalLanguageResponse,
    IntentType
)

class NaturalLanguageTool:
    def __init__(self, airtable_service, logger: Optional[logging.Logger] = None):
        self.service = airtable_service
        self.logger = logger or logging.getLogger(__name__)
        
        # Initialize NLP processor
        nlp_config = NLPConfig(
            confidence_threshold=0.5,
            max_context_queries=10,
            enable_date_processing=True,
            enable_contextual_references=True,
            supported_languages=['es', 'en'],
            synonyms={
                'crear': ['nuevo', 'añadir', 'agregar', 'insertar'],
                'mostrar': ['listar', 'ver', 'obtener', 'mostrarme'],
                'actualizar': ['modificar', 'cambiar', 'editar'],
                'eliminar': ['borrar', 'suprimir', 'remover'],
                'buscar': ['filtrar', 'encontrar', 'buscar']
            },
            priority_synonyms={
                'alta': 'Alta',
                'media': 'Media',
                'baja': 'Baja',
                'alta prioridad': 'Alta',
                'baja prioridad': 'Baja'
            }
        )
        
        self.processor = NaturalLanguageProcessor(nlp_config, self.logger)

    async def process_natural_language_query(
        self, 
        query: str, 
        session_id: str = 'default', 
        user_id: str = 'anonymous'
    ) -> Dict[str, Any]:
        """
        Main method to process natural language queries
        """
        start_time = self.logger.info(f"Processing NLP query: {query}")
        
        try:
            # Create natural language query
            nl_query = NaturalLanguageQuery(
                query=query.strip(),
                session_id=session_id,
                user_id=user_id
            )

            # Process the query
            processed_query = await self.processor.process_query(nl_query)
            
            # Check if clarification is needed
            if processed_query.requires_clarification:
                response = NaturalLanguageResponse(
                    success=False,
                    message="Necesito más información para procesar tu consulta.",
                    intent=processed_query.intent,
                    confidence=processed_query.confidence,
                    clarifications=processed_query.clarifications,
                    metadata={
                        'processing_time': 0,  # Will be calculated
                        'context_used': True,
                        'fallback_used': False
                    }
                )
                return self._response_to_dict(response)

            # Execute the corresponding tool
            tool_result = await self._execute_corresponding_tool(processed_query)
            
            # Create natural language response
            response = NaturalLanguageResponse(
                success=True,
                data=tool_result,
                message=self._generate_natural_language_response(
                    processed_query.intent, tool_result, query
                ),
                intent=processed_query.intent,
                confidence=processed_query.confidence,
                metadata={
                    'processing_time': 0,  # Will be calculated
                    'context_used': True,
                    'fallback_used': False
                }
            )

            return self._response_to_dict(response)

        except Exception as e:
            self.logger.error(f"Error processing natural language query: {e}")
            
            error_response = NaturalLanguageResponse(
                success=False,
                message=f"Lo siento, ocurrió un error al procesar tu consulta: {str(e)}",
                intent=IntentType.UNKNOWN,
                confidence=0.0,
                metadata={
                    'processing_time': 0,
                    'context_used': False,
                    'fallback_used': True
                }
            )
            return self._response_to_dict(error_response)

    async def _execute_corresponding_tool(self, processed_query: ProcessedQuery) -> Dict[str, Any]:
        """Execute the tool corresponding to the processed intent"""
        intent = processed_query.intent
        parameters = processed_query.parameters

        # Map intent to tool execution
        try:
            if intent == IntentType.LIST_BASES:
                result = await self.service.list_bases()
                return {'bases': result.get('bases', [])}
            
            elif intent == IntentType.LIST_RECORDS:
                if not parameters.table:
                    raise ValueError("Nombre de tabla requerido")
                result = await self.service.list_records(
                    base_id=parameters.base_id or 'default',
                    table_name=parameters.table,
                    filter_by_formula=parameters.filter_by_formula,
                    max_records=parameters.max_records
                )
                return result
            
            elif intent == IntentType.CREATE_RECORD:
                if not parameters.table or not parameters.fields:
                    raise ValueError("Tabla y campos requeridos")
                result = await self.service.create_record(
                    base_id=parameters.base_id or 'default',
                    table_name=parameters.table,
                    fields=parameters.fields
                )
                return result
            
            elif intent == IntentType.UPDATE_RECORD:
                if not parameters.table or not parameters.record_id or not parameters.fields:
                    raise ValueError("Tabla, ID de registro y campos requeridos")
                result = await self.service.update_record(
                    base_id=parameters.base_id or 'default',
                    table_name=parameters.table,
                    record_id=parameters.record_id,
                    fields=parameters.fields
                )
                return result
            
            elif intent == IntentType.DELETE_RECORD:
                if not parameters.table or not parameters.record_id:
                    raise ValueError("Tabla e ID de registro requeridos")
                result = await self.service.delete_record(
                    base_id=parameters.base_id or 'default',
                    table_name=parameters.table,
                    record_id=parameters.record_id
                )
                return result
            
            elif intent == IntentType.SEARCH_RECORDS:
                if not parameters.table:
                    raise ValueError("Nombre de tabla requerido")
                result = await self.service.list_records(
                    base_id=parameters.base_id or 'default',
                    table_name=parameters.table,
                    filter_by_formula=parameters.filter_by_formula,
                    max_records=parameters.max_records
                )
                return result
            
            elif intent == IntentType.LIST_TABLES:
                result = await self.service.list_tables(
                    base_id=parameters.base_id or 'default'
                )
                return result
            
            elif intent == IntentType.CREATE_WEBHOOK:
                if not parameters.table or not parameters.webhook_config:
                    raise ValueError("Tabla y configuración de webhook requeridas")
                result = await self.service.create_webhook(
                    base_id=parameters.base_id or 'default',
                    table_name=parameters.table,
                    webhook_config=parameters.webhook_config
                )
                return result
            
            elif intent == IntentType.LIST_WEBHOOKS:
                result = await self.service.list_webhooks(
                    base_id=parameters.base_id or 'default'
                )
                return result
            
            elif intent == IntentType.GET_BASE_SCHEMA:
                if not parameters.base_id:
                    raise ValueError("ID de base requerido")
                result = await self.service.get_base_schema(
                    base_id=parameters.base_id
                )
                return result
            
            elif intent == IntentType.DESCRIBE_TABLE:
                if not parameters.table:
                    raise ValueError("Nombre de tabla requerido")
                result = await self.service.describe_table(
                    base_id=parameters.base_id or 'default',
                    table_name=parameters.table
                )
                return result
            
            else:
                # For intents not yet implemented, return a placeholder response
                return {
                    'intent': intent.value,
                    'parameters': parameters.__dict__,
                    'message': f'Operación {intent.value} solicitada pero no implementada completamente',
                    'timestamp': '2024-01-01T00:00:00Z'
                }
        
        except Exception as e:
            self.logger.error(f"Error executing tool for intent {intent}: {e}")
            raise

    def _generate_natural_language_response(
        self, 
        intent: IntentType, 
        result: Dict[str, Any], 
        original_query: str
    ) -> str:
        """Generate natural language response for the operation"""
        responses = {
            IntentType.LIST_BASES: 'He encontrado las siguientes bases Airtable accesibles:',
            IntentType.LIST_RECORDS: 'Aquí están los registros encontrados en la tabla:',
            IntentType.CREATE_RECORD: 'He creado el nuevo registro exitosamente.',
            IntentType.UPDATE_RECORD: 'He actualizado el registro como solicitaste.',
            IntentType.DELETE_RECORD: 'He eliminado el registro especificado.',
            IntentType.SEARCH_RECORDS: 'Aquí están los resultados de tu búsqueda:',
            IntentType.LIST_TABLES: 'Las tablas disponibles en tu base son:',
            IntentType.CREATE_WEBHOOK: 'He creado el webhook correctamente.',
            IntentType.LIST_WEBHOOKS: 'Los webhooks activos en tu base son:',
            IntentType.GET_BASE_SCHEMA: 'El esquema completo de tu base es:',
            IntentType.DESCRIBE_TABLE: 'Los detalles de la tabla solicitada son:',
            IntentType.UNKNOWN: 'No pude entender completamente tu consulta.'
        }

        base_response = responses.get(intent, 'Operación completada.')
        
        # Add specific details based on result
        if intent == IntentType.LIST_BASES and 'bases' in result:
            base_count = len(result['bases'])
            return f"{base_response} Encontré {base_count} bases accesibles."
        
        elif intent == IntentType.LIST_RECORDS and 'records' in result:
            record_count = len(result['records'])
            return f"{base_response} Se encontraron {record_count} registros."
        
        return base_response

    def _response_to_dict(self, response: NaturalLanguageResponse) -> Dict[str, Any]:
        """Convert NaturalLanguageResponse to dictionary"""
        return {
            'success': response.success,
            'data': response.data,
            'message': response.message,
            'intent': response.intent.value,
            'confidence': response.confidence,
            'clarifications': [
                {
                    'question': c.question,
                    'type': c.type.value,
                    'suggestions': c.suggestions,
                    'required': c.required
                }
                for c in (response.clarifications or [])
            ],
            'metadata': response.metadata
        }

    async def get_context_summary(self, session_id: str) -> Dict[str, Any]:
        """Get summary of current conversation context"""
        return await self.processor.get_context_summary(session_id)

    async def clear_context(self, session_id: str) -> None:
        """Clear conversation context"""
        await self.processor.clear_context(session_id)

# MCP Tool Registration Functions
def register_nlp_tools(mcp, service):
    """Register all NLP tools with MCP server"""
    nlp_tool = NaturalLanguageTool(service)
    
    @mcp.tool()
    async def process_natural_language(query: str, session_id: str = 'default', user_id: str = 'anonymous') -> str:
        """Procesa consultas en lenguaje natural español y las ejecuta como operaciones de Airtable"""
        result = await nlp_tool.process_natural_language_query(query, session_id, user_id)
        return json.dumps(result, indent=2, ensure_ascii=False)

    @mcp.tool()
    async def get_nlp_context(session_id: str) -> str:
        """Obtiene el contexto actual de la conversación"""
        context = await nlp_tool.get_context_summary(session_id)
        return json.dumps({
            'success': True,
            'data': context,
            'message': 'Contexto obtenido correctamente'
        }, indent=2, ensure_ascii=False)

    @mcp.tool()
    async def clear_nlp_context(session_id: str) -> str:
        """Limpia el contexto conversacional de una sesión"""
        await nlp_tool.clear_context(session_id)
        return json.dumps({
            'success': True,
            'message': 'Contexto limpiado correctamente'
        }, indent=2, ensure_ascii=False)

File: src/typescript/app/tools/natural-language.ts
/**
 * Herramienta MCP para procesamiento de consultas en lenguaje natural
 */

import { McpServer } from '@modelcontextprotocol/sdk/server/mcp';
import { AppContext } from '../context';
import { 
  NaturalLanguageProcessor, 
  NLPConfig,
  NaturalLanguageQuery,
  ProcessedQuery,
  NaturalLanguageResponse,
  IntentType
} from '../nlp';
import { Logger } from '../logger';

export function registerNaturalLanguageTool(server: McpServer, ctx: AppContext): void {
  const nlpConfig: NLPConfig = {
    confidenceThreshold: 0.5,
    maxContextQueries: 10,
    enableDateProcessing: true,
    enableContextualReferences: true,
    supportedLanguages: ['es', 'en'],
    synonyms: {
      'crear': ['nuevo', 'añadir', 'agregar', 'insertar'],
      'mostrar': ['listar', 'ver', 'obtener', 'mostrarme'],
      'actualizar': ['modificar', 'cambiar', 'editar'],
      'eliminar': ['borrar', 'suprimir', 'remover'],
      'buscar': ['filtrar', 'encontrar', 'buscar']
    },
    entityPatterns: {},
    datePatterns: [],
    prioritySynonyms: {
      'alta': 'Alta',
      'media': 'Media', 
      'baja': 'Baja',
      'alta prioridad': 'Alta',
      'baja prioridad': 'Baja'
    }
  };

  const processor = new NaturalLanguageProcessor(nlpConfig, ctx.logger);

  // Herramienta principal para procesar consultas en lenguaje natural
  server.setRequestHandler('tools/call', async (request) => {
    const { name, arguments: args } = request.params;

    if (name === 'process_natural_language') {
      return await handleNaturalLanguageQuery(args, processor, ctx);
    }

    if (name === 'get_nlp_context') {
      return await handleGetContext(args, ctx);
    }

    if (name === 'clear_nlp_context') {
      return await handleClearContext(args, ctx);
    }

    throw new Error(`Herramienta desconocida: ${name}`);
  });

  // Registrar las herramientas disponibles
  server.setRequestHandler('tools/list', async () => {
    return {
      tools: [
        {
          name: 'process_natural_language',
          description: 'Procesa consultas en lenguaje natural español y las ejecuta como operaciones de Airtable',
          inputSchema: {
            type: 'object',
            properties: {
              query: {
                type: 'string',
                description: 'La consulta en lenguaje natural a procesar'
              },
              sessionId: {
                type: 'string',
                description: 'ID de sesión para mantener contexto conversacional'
              },
              userId: {
                type: 'string',
                description: 'ID del usuario para personalizar la experiencia'
              }
            },
            required: ['query']
          }
        },
        {
          name: 'get_nlp_context',
          description: 'Obtiene el contexto actual de la conversación',
          inputSchema: {
            type: 'object',
            properties: {
              sessionId: {
                type: 'string',
                description: 'ID de sesión para obtener el contexto'
              }
            },
            required: ['sessionId']
          }
        },
        {
          name: 'clear_nlp_context',
          description: 'Limpia el contexto conversacional de una sesión',
          inputSchema: {
            type: 'object',
            properties: {
              sessionId: {
                type: 'string',
                description: 'ID de sesión para limpiar el contexto'
              }
            },
            required: ['sessionId']
          }
        }
      ]
    };
  });
}

async function handleNaturalLanguageQuery(
  args: any, 
  processor: NaturalLanguageProcessor, 
  ctx: AppContext
) {
  const { query, sessionId = 'default', userId = 'anonymous' } = args;

  if (!query || typeof query !== 'string') {
    throw new Error('Se requiere una consulta válida');
  }

  const startTime = Date.now();
  
  try {
    // Crear consulta de lenguaje natural
    const nlQuery: NaturalLanguageQuery = {
      query: query.trim(),
      sessionId,
      userId
    };

    // Procesar la consulta
    const processedQuery: ProcessedQuery = await processor.processQuery(nlQuery);
    
    // Verificar si necesita clarificación
    if (processedQuery.requiresClarification) {
      const response: NaturalLanguageResponse = {
        success: false,
        message: 'Necesito más información para procesar tu consulta.',
        intent: processedQuery.intent,
        confidence: processedQuery.confidence,
        clarifications: processedQuery.clarifications,
        metadata: {
          processingTime: Date.now() - startTime,
          contextUsed: true,
          fallbackUsed: false
        }
      };

      return {
        content: [{
          type: 'text',
          text: JSON.stringify(response, null, 2)
        }]
      };
    }

    // Ejecutar la herramienta correspondiente
    const toolResult = await executeCorrespondingTool(processedQuery, ctx);
    
    // Crear respuesta en lenguaje natural
    const response: NaturalLanguageResponse = {
      success: true,
      data: toolResult,
      message: generateNaturalLanguageResponse(processedQuery.intent, toolResult, query),
      intent: processedQuery.intent,
      confidence: processedQuery.confidence,
      metadata: {
        processingTime: Date.now() - startTime,
        contextUsed: true,
        fallbackUsed: false
      }
    };

    return {
      content: [{
        type: 'text',
        text: JSON.stringify(response, null, 2)
      }]
    };

  } catch (error) {
    ctx.logger.error('Error processing natural language query', { error, query });
    
    const errorResponse: NaturalLanguageResponse = {
      success: false,
      message: `Lo siento, ocurrió un error al procesar tu consulta: ${error instanceof Error ? error.message : 'Error desconocido'}`,
      intent: 'unknown',
      confidence: 0,
      metadata: {
        processingTime: Date.now() - startTime,
        contextUsed: false,
        fallbackUsed: true
      }
    };

    return {
      content: [{
        type: 'text',
        text: JSON.stringify(errorResponse, null, 2)
      }]
    };
  }
}

async function handleGetContext(args: any, ctx: AppContext) {
  const { sessionId = 'default' } = args;
  
  try {
    const contextHandler = new (await import('../nlp/context-handler')).ContextHandler();
    const context = await contextHandler.getContext(sessionId, 'anonymous');
    
    return {
      content: [{
        type: 'text',
        text: JSON.stringify({
          success: true,
          data: context,
          message: 'Contexto obtenido correctamente'
        }, null, 2)
      }]
    };
  } catch (error) {
    return {
      content: [{
        type: 'text',
        text: JSON.stringify({
          success: false,
          message: `Error al obtener contexto: ${error instanceof Error ? error.message : 'Error desconocido'}`
        }, null, 2)
      }]
    };
  }
}

async function handleClearContext(args: any, ctx: AppContext) {
  const { sessionId = 'default' } = args;
  
  try {
    const contextHandler = new (await import('../nlp/context-handler')).ContextHandler();
    await contextHandler.clearContext(sessionId);
    
    return {
      content: [{
        type: 'text',
        text: JSON.stringify({
          success: true,
          message: 'Contexto limpiado correctamente'
        }, null, 2)
      }]
    };
  } catch (error) {
    return {
      content: [{
        type: 'text',
        text: JSON.stringify({
          success: false,
          message: `Error al limpiar contexto: ${error instanceof Error ? error.message : 'Error desconocido'}`
        }, null, 2)
      }]
    };
  }
}

async function executeCorrespondingTool(processedQuery: ProcessedQuery, ctx: AppContext): Promise<any> {
  const intent = processedQuery.intent;
  const parameters = processedQuery.parameters;

  // Mapear intención a herramienta específica
  const toolMapping: Record<IntentType, string> = {
    'list_bases': 'list_bases',
    'list_records': 'list_records',
    'create_record': 'create_record',
    'update_record': 'update_record',
    'delete_record': 'delete_record',
    'search_records': 'search_records',
    'list_tables': 'list_tables',
    'get_record': 'get_record',
    'create_webhook': 'create_webhook',
    'list_webhooks': 'list_webhooks',
    'delete_webhook': 'delete_webhook',
    'get_webhook_payloads': 'get_webhook_payloads',
    'get_base_schema': 'get_base_schema',
    'describe_table': 'describe_table',
    'create_table': 'create_table',
    'delete_table': 'delete_table',
    'update_table': 'update_table',
    'create_field': 'create_field',
    'delete_field': 'delete_field',
    'update_field': 'update_field',
    'list_field_types': 'list_field_types',
    'batch_create_records': 'batch_create_records',
    'batch_update_records': 'batch_update_records',
    'batch_delete_records': 'batch_delete_records',
    'batch_upsert_records': 'batch_upsert_records',
    'upload_attachment': 'upload_attachment',
    'list_collaborators': 'list_collaborators',
    'list_shares': 'list_shares',
    'create_view': 'create_view',
    'get_view_metadata': 'get_view_metadata',
    'get_table_views': 'get_table_views',
    'create_base': 'create_base',
    'analyze_data': 'analyze_data',
    'create_report': 'create_report',
    'data_insights': 'data_insights',
    'optimize_workflow': 'optimize_workflow',
    'smart_schema_design': 'smart_schema_design',
    'data_quality_audit': 'data_quality_audit',
    'predictive_analytics': 'predictive_analytics',
    'natural_language_query': 'natural_language_query',
    'smart_data_transformation': 'smart_data_transformation',
    'automation_recommendations': 'automation_recommendations',
    'unknown': 'unknown'
  };

  const toolName = toolMapping[intent];
  
  if (toolName === 'unknown') {
    throw new Error(`Intención no reconocida: ${intent}`);
  }

  // Aquí se integraría con el sistema de herramientas existente
  // Por simplicidad, simulamos la ejecución
  return {
    intent,
    parameters,
    message: `Herramienta ${toolName} ejecutada con éxito`,
    timestamp: new Date().toISOString()
  };
}

function generateNaturalLanguageResponse(intent: IntentType, result: any, originalQuery: string): string {
  const responses: Record<IntentType, string> = {
    'list_bases': 'He encontrado las siguientes bases Airtable accesibles:',
    'list_records': 'Aquí están los registros encontrados en la tabla:',
    'create_record': 'He creado el nuevo registro exitosamente.',
    'update_record': 'He actualizado el registro como solicitaste.',
    'delete_record': 'He eliminado el registro especificado.',
    'search_records': 'Aquí están los resultados de tu búsqueda:',
    'list_tables': 'Las tablas disponibles en tu base son:',
    'get_record': 'Aquí está la información del registro solicitado:',
    'create_webhook': 'He creado el webhook correctamente.',
    'list_webhooks': 'Los webhooks activos en tu base son:',
    'delete_webhook': 'He eliminado el webhook especificado.',
    'get_webhook_payloads': 'Aquí están las notificaciones recientes del webhook:',
    'get_base_schema': 'El esquema completo de tu base es:',
    'describe_table': 'Los detalles de la tabla solicitada son:',
    'create_table': 'He creado la nueva tabla exitosamente.',
    'delete_table': 'He eliminado la tabla especificada.',
    'update_table': 'He actualizado la tabla como solicitaste.',
    'create_field': 'He agregado el nuevo campo a la tabla.',
    'delete_field': 'He eliminado el campo especificado.',
    'update_field': 'He actualizado el campo como solicitaste.',
    'list_field_types': 'Los tipos de campos disponibles en Airtable son:',
    'batch_create_records': 'He creado los registros en lote exitosamente.',
    'batch_update_records': 'He actualizado los registros en lote como solicitaste.',
    'batch_delete_records': 'He eliminado los registros en lote especificados.',
    'batch_upsert_records': 'He realizado las operaciones de actualización/creación en lote.',
    'upload_attachment': 'He adjuntado el archivo al registro.',
    'list_collaborators': 'Los colaboradores de esta base son:',
    'list_shares': 'Las vistas compartidas en esta base son:',
    'create_view': 'He creado la nueva vista exitosamente.',
    'get_view_metadata': 'Los metadatos de la vista solicitada son:',
    'get_table_views': 'Las vistas de la tabla son:',
    'create_base': 'He creado la nueva base exitosamente.',
    'analyze_data': 'He completado el análisis de datos solicitado.',
    'create_report': 'He generado el informe solicitado.',
    'data_insights': 'Aquí están las perspectivas de los datos:',
    'optimize_workflow': 'Las recomendaciones de optimización son:',
    'smart_schema_design': 'El diseño de esquema inteligente sugiere:',
    'data_quality_audit': 'Los resultados de la auditoría de calidad son:',
    'predictive_analytics': 'Los resultados del análisis predictivo son:',
    'natural_language_query': 'He procesado tu consulta en lenguaje natural.',
    'smart_data_transformation': 'La transformación de datos inteligente está completa.',
    'automation_recommendations': 'Las recomendaciones de automatización son:',
    'unknown': 'No pude entender completamente tu consulta.'
  };

  return responses[intent] || 'Operación completada.';
}
